### ISSUE TITLE
Add info when a grammar rule fails but it was clearly meant to pass for the user

### ISSUE THREAD

**Issue by BrunoC-L** [BrunoC-L/Caesium#1]
Say user code contains
```
type Cat:
    String ref name
```
the rule for type will fail, because the rule for class member will fail, because none of its rules will match `String ref name`
The user made a mistake and type members can't be references, they should have used `String name`.
As of now, there is no diagnostic, but ideally when we begin parsing type members, there should be a diagnostic such as `String ref name` is not a valid class element.
The basis for doing so is that we can tell this rule *should* work because of the parsing context inside a type.
The rule for parsing types would have some capability to know that once it begins parsing it has the authority to produce an error rather than returning false if it is unable to complete.
Basically, there are no rules other than the rule for types that will accept `type` so once we see that token it is 100% sure that a type is what the user is trying to produce here.
